lnwallet/channel_test.go

```diff
@@ -4824,24 +4846,53 @@ func TestChanAvailableBalanceNearHtlcFee(t *testing.T) {
 	// nonDustHtlc+commitfee+aliceReserve+htlcfee. The available balance
 	// reported will just be nonDustHtlc, since the rest of the balance is
 	// reserved.
-	expBalance = aliceNonDustHtlc
-	checkBalance(t, expBalance)
+	expAliceBalance = aliceNonDustHtlc
+	expBobBalance = bobBalance - bobReserve
+	checkBalance(t, expAliceBalance, expBobBalance)
 
 	// Send an HTLC using all but one msat of the reported balance.
 	htlcAmt = aliceNonDustHtlc - 1
 	sendHtlc(htlcAmt)
 
 	// 1 msat should be left.
-	expBalance = 1
-	checkBalance(t, expBalance)
+	expAliceBalance = 1
+
+	// Bob should still have all his balance available, since even though
+	// Alice cannot afford to add a non-dust HTLC, she can afford to add a
+	// non-dust HTLC from Bob.
+	expBobBalance = bobBalance - bobReserve
+	checkBalance(t, expAliceBalance, expBobBalance)
 
 	// Sendng the last msat.
 	htlcAmt = 1
 	sendHtlc(htlcAmt)
 
 	// No balance left.
-	expBalance = 0
-	checkBalance(t, expBalance)
+	expAliceBalance = 0
+
+	// We try to always reserve enough for the non-iniitator to be able to
+	// add an HTLC, hence Bob should still have all his non-reserved
+	// balance available.
+	expBobBalance = bobBalance - bobReserve
+	checkBalance(t, expAliceBalance, expBobBalance)
+
+	// Even though Alice has a reported balance of 0, this is because we
+	// try to avoid getting into the position where she cannot pay the fee
+	// for Bob adding another HTLC. This means she actually _has_ some
+	// balance left, and we now force the channel into this situation by
+	// sending yet another HTLC. In practice this can also happen if a fee
+	// update eats into Alice's balance.
+	htlcAmt = 1
+	sendHtlc(htlcAmt)
```

Comment:
Hmm is it good that this is still possible if there is no amount sendable anymore?

Reply:
It would be changing `AddHTLC` to not allow sending HTLCs even though there is balance left. IMO it is a nice upper layer heuristic to not attempt this by reporting this balance not being available. We can revisit if this becomes a hard protocol requirement,  re https://github.com/lightningnetwork/lightning-rfc/issues/728

Reply:
It could lead to getting caught in the danger zone if there are concurrent payments. Both payments used the same reported `availableBalance`, which was enough at that point, but when executed the second payment pushes us below the threshold. If `AddHTLC` would fail, this wouldn't happen and a new path for the second payment could be searched for.

But this is arguably a rare condition, especially because we plan to do non-concurrent pathfinding for mpp. Maybe it is worth a comment near the `AddHTLC` implementation.

Reply:
Btw, does this mean that you can still go below the threshold with SendToRoute?

Reply:
`SendToRoute` will still eventually send an HTLC through the link, so none of these checks are bypassed.

Reply:
`SendToRoute` will also be forwarded by the switch, and `Bandwidth` is called here: https://github.com/lightningnetwork/lnd/blob/92b79f6b6abc7e027ee1d7253b7c4e67c0090476/htlcswitch/link.go#L2339